policy: clean up synthesised deny entries

bimmlerd opened the pull request:
When there is an overlap in policies a case can arise where cilium needs to synthesise map state entries for specific identities to avoid punching holes in broad deny policies. One such example is:

egressDeny toPorts TCP 123
toFQDN matchPattern: *

When cilium allocates a new (locally-scoped) identity for an IP learned via DNS interception and inserts it into the policy map of an EP, it avoids punching a hole into the deny by inserting "synthesised", more specific deny entries for the new identity and TCP port 123.

This works, but the implementation neglects to handle the case that, uniquely, FQDN selectors can stop selecting the identity _without policy changing_. The incremental deletion path only removes the newly added allow entry, without cleaning up the synthesised denies. It's possible to write policies which blow up the number of synthesised entries by a significant factor - combined with the identity churn of FQDN identities, policy map space is exhausted quickly. Any policy revision bump or other reasons for policy recalculation fix this, but aren't necessarily common.

The actual change is somewhat subtle: instead of the broad deny entry the new allow owns the more specific deny copy. The key insight is that a wildcarded identity selector can not incrementally deselect an identity like the FQDN selector can.[1] We can thus use the ownership/dependency infrastructure to have the deny copies be cleaned up when the allow is removed.

Since we're in the specific branch where we're inserting an allow which has a more specific identity, but a less specific port/proto, the new entry is always the right choice of owner.

[1] Note that this applies only to 1.15 and older - 1.16 has transformed the FQDN selectors to also not change selections dynamically.

```release-note
Cilium avoids running out of space in policy maps by cleaning up entries in specific cases previously missed.
```

bimmlerd committed:
policy: clean up synthesized deny entries

When there is an overlap in policies a case can arise where cilium needs
to synthesise map state entries for specific identities to avoid
punching holes in broad deny policies. One such example is:

egressDeny toPorts TCP 123
toFQDN matchPattern: *

When cilium allocates a new (locally-scoped) identity for an IP learned
via DNS interception and inserts it into the policy map of an EP, it
avoids punching a hole into the deny by inserting "synthesised", more
specific deny entries for the new identity and TCP port 123.

This works, but the implementation neglects to handle the case that,
uniquely, FQDN selectors can stop selecting the identity _without policy
changing_. The incremental deletion path only removes the newly added
allow entry, without cleaning up the synthesised denies. It's possible
to write policies which blow up the number of synthesised entries by a
significant factor - combined with the identity churn of FQDN
identities, policy map space is exhausted quickly. Any policy revision
bump or other reasons for policy recalculation fix this, but aren't
necessarily common.

The actual change is somewhat subtle: instead of the broad deny entry
the new allow owns the more specific deny copy. The key insight is that
a wildcarded identity selector can not incrementally deselect an
identity like the FQDN selector can.[1] We can thus use the
ownership/dependency infrastructure to have the deny copies be cleaned
up when the allow is removed.

Since we're in the specific branch where we're inserting an allow which
has a more specific identity, but a less specific port/proto, the new
entry is always the right choice of owner.

[1] Note that this applies only to 1.15 and older - 1.16 has transformed
the FQDN selectors to also not change selections dynamically.

Suggested-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

jrajahalme commented on pkg/policy/mapstate.go:
```diff
@@ -933,12 +933,14 @@ func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapState
-					l3l4DenyEntry := NewMapStateEntry(k, v.DerivedFromRules, false, true, DefaultAuthType, AuthTypeDisabled)
+					// Note that we let the new key own this copy, since only
+					// the new key can be incrementally deleted.
+					l3l4DenyEntry := NewMapStateEntry(newKey, v.DerivedFromRules, false, true, DefaultAuthType, AuthTypeDisabled)
```
I think there are other instances above where the owner of the synthetic entry is a deny key or a key with `0` identity. Should probably flip the ownership of those as well in the same way.

jrajahalme replied:
Basically go through each line that adds a synthetic entry and ask if the first argument of `NewMapStateEntry` either:
- has a `0` identity, or
- is a deny key

If yes flip the ownership.

bimmlerd replied:
I tried to apply this, but it seemed to only apply to one other case.

If both k and newKey was deny/had zero identity I think it doesn't actually matter which owns the synthesized entry, so I left those.

jrajahalme reviewed (changes requested):
It looks to me there are more than this one instance to fix.

bimmlerd committed:
policy: swap owner of synthesized map state entry

Having deny keys own synthesised map state entries does not make sense,
because denies cannot be incrementally removed.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
policy: add test for clean up of synthesized entry

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

christarazi commented on pkg/policy/mapstate_test.go:
```diff
@@ -1426,6 +1426,42 @@ func (ds *PolicyTestSuite) TestMapState_AccumulateMapChangesDeny(c *check.C) {
+		}),
+		args: []args{
+			// Simulate addition of toFQDN identity
+			{cs: csFoo, adds: []int{16777218}, deletes: []int{}, port: 80, proto: 6, ingress: false, redirect: false, deny: true},
+			{cs: csBar, adds: []int{16777218}, deletes: []int{}, port: 0, proto: 0, ingress: false, redirect: false, deny: false},
```
Does it make a difference / help coverage if we tested with a different identity as well?

bimmlerd replied:
not sure what you mean - I've chosen this identity for being locally scoped because that was the case I had in mind, but I'm not sure it matters.
By different identity you mean an additional identity or a different numeric identity?

christarazi replied:
Sorry, I meant a different numeric identity.

bimmlerd replied:
I don't believe so, the mapstate seems mostly angostic to whether an ID is locally scoped or not.

jrajahalme reviewed (changes requested):
Two more call sites to check out
